🏷️ feat: Request Body Placeholders in Custom Headers

s10gopal opened the pull request:
## Summary
feat: add support for request body placeholders in custom endpoint headers

- Add {{LIBRECHAT_BODY_*}} placeholders for conversationId, parentMessageId, messageId

## Change Type

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Testing

Please describe your test process and include instructions so that we can reproduce your test. If there are any important variables for your testing configuration, list them here.

### **Test Configuration**:

endpoints:
  custom:
    - name: 'customllm'
      apiKey: 'KEY'
      baseURL: 'http://host.docker.internal:8001/v1'
      models:
        default:
          [
            'mycustommodel'
          ]
        fetch: false
      titleConvo: true
      titleModel: 'current_model'
      modelDisplayLabel: 'mycustommodel'
      headers:
        X-Conversation-ID: "{{LIBRECHAT_BODY_CONVERSATIONID}}"
        
## Checklist

- [x] My code adheres to this project's style guidelines
- [x] I have performed a self-review of my own code
- [x] I have made pertinent documentation changes
- [x] My changes do not introduce new warnings
- [x] I have written tests demonstrating that my changes are effective or that my feature works
- [x] Local unit tests pass with my changes
- [x] A pull request for updating the documentation has been submitted. https://github.com/LibreChat-AI/librechat.ai/pull/386

Commit:
feat: Add conversation ID support to custom endpoint headers

- Add LIBRECHAT_CONVERSATION_ID to customUserVars when provided
- Pass conversation ID to header resolution for dynamic headers
- Add comprehensive test coverage

Enables custom endpoints to access conversation context using {{LIBRECHAT_CONVERSATION_ID}} placeholder.

MrunmayS commented on api/server/services/Endpoints/custom/initialize.js:
```diff
@@ -28,7 +28,10 @@ const initializeClient = async ({ req, res, endpointOption, optionsOnly, overrid
 
-  let resolvedHeaders = resolveHeaders(endpointConfig.headers, req.user);
+  const customUserVars = {};
+  customUserVars.LIBRECHAT_CONVERSATION_ID = req.body.conversationId;
```
I was working on a similar change `conversationId`
For the First Request in a new thread there would be no `consersationId`. We need to filter out the unresolved placeholder for `conversationId`. Otherwise `customUserVars { LIBRECHAT_CONVERSATION_ID: undefined }` for the first call

```js
  const customUserVars = {};
  if (req.body.conversationId) {
    customUserVars.LIBRECHAT_CONVERSATION_ID = req.body.conversationId;
  }

  let resolvedHeaders = resolveHeaders(endpointConfig.headers, req.user, customUserVars);

  // Filter out headers with unresolved placeholders
  const filteredHeaders = {};
  for (const [key, value] of Object.entries(resolvedHeaders)) {
    if (typeof value === 'string' && value.includes('{{') && value.includes('}}')) {
      continue;
    }
    filteredHeaders[key] = value;
  }

  resolvedHeaders = filteredHeaders;
```

s10gopal replied:
Thanks! I've incorporated the suggested code to filter out unresolved placeholders from the headers.

s10gopal committed:
fix: filter out unresolved placeholders from headers (thanks @MrunmayS)

MrunmayS reviewed (commented):
Looks Good!

s10gopal commented:
Hi @danny-avila, Could you please review?

danny-avila commented on api/server/services/Endpoints/custom/initialize.js:
```diff
@@ -28,7 +28,23 @@ const initializeClient = async ({ req, res, endpointOption, optionsOnly, overrid
   const CUSTOM_BASE_URL = extractEnvVariable(endpointConfig.baseURL);
 
-  let resolvedHeaders = resolveHeaders(endpointConfig.headers, req.user);
+  const customUserVars = {};
```
This only applies this change to custom endpoints and should not be directly in this module.

Changes should be made here: `packages/api/src/utils/env.ts`

And tests added here: `/packages/api/src/utils/env.spec.ts`

s10gopal replied:
To support request-scoped placeholders like {{LIBRECHAT_USER_CONVERSATIONID}}, {{LIBRECHAT_USER_MESSAGEID}} in env.ts, possible options could be:

- Pass `req.body` into the relevant env utility functions (`resolveHeaders`, `processSingleValue`, etc.) to access values like `req.body.conversationId`.

- Create a helper function like `extractRequestScopedVars(endpointConfig.headers,req.body)` to return a customUserVars object, which can then be passed to resolveHeaders.

Would you recommend either of these, or suggest a different approach?

Thanks

Commit:
feat: add support for request body placeholders in custom endpoint headers

- Add {{LIBRECHAT_BODY_*}} placeholders for conversationId, parentMessageId, messageId
- Update tests to reflect new body placeholder functionality

s10gopal reviewed (commented):
Thanks for the guidance! I've moved the logic to env.ts and added tests in env.spec.ts. Happy to adjust further if needed

danny-avila commented on packages/api/src/utils/env.ts:
```diff
@@ -61,21 +71,46 @@ function processUserPlaceholders(value: string, user?: TUser): string {
   originalValue: string;
   customUserVars?: Record<string, string>;
   user?: TUser;
+  body?: Record<string, any>;
```
`any` type is not allowed, please type `body` correctly

danny-avila commented on packages/api/src/utils/env.ts:
```diff
@@ -151,16 +191,18 @@ export function processMCPEnv(
   headers: Record<string, string> | undefined,
   user?: Partial<TUser> | { id: string },
   customUserVars?: Record<string, string>,
+  body?: Record<string, any>,
```
`any` type is not allowed, please type `body` correctly

dustinhealy commented:
I'll continue on from where you guys left off here and resolve the rest of Danny's requested changes. Thanks for the work done so far!

s10gopal commented:
> I'll continue on from where you guys left off here and resolve the rest of Danny's requested changes. Thanks for the work done so far!

Thanks! I’m working on those changes now and will push within the hour.

Commit:
refactor resolveHeaders

s10gopal commented:
@dustinhealy, Changes pushed. Please review when you can.

Thanks

s10gopal commented:
Docs PR: https://github.com/LibreChat-AI/librechat.ai/pull/386

dustinhealy committed:
style: minor styling cleanup

dustinhealy committed:
fix: type error in unit test

s10gopal committed:
Merge branch 'dev' into feat/custom-endpoint-conversation-id

dustinhealy committed:
feat: add body to other endpoints

dustinhealy committed:
feat: add body for mcp tool calls

s10gopal committed:
Merge branch 'dev' into feat/custom-endpoint-conversation-id

s10gopal committed:
Merge branch 'dev' into feat/custom-endpoint-conversation-id

s10gopal committed:
Merge branch 'dev' into feat/custom-endpoint-conversation-id

dustinhealy committed:
Merge remote-tracking branch 'origin/dev' into pr/s10gopal/8601

dustinhealy committed:
chore: remove changes that unnecessarily increase scope after clarification of requirements

s10gopal commented:
Hi @danny-avila, Could you please review?

s10gopal committed:
Merge branch 'dev' into feat/custom-endpoint-conversation-id

dustinhealy committed:
refactor: move http.ts to packages/api and have RequestBody intersect with Express request body

s10gopal commented:
Hi @danny-avila, really appreciate your earlier reviews and guidance. When you have a moment, could you take another look?

dustinhealy committed:
refactor: processMCPEnv now uses single object argument pattern

danny-avila commented:
Moved to https://github.com/danny-avila/LibreChat/pull/9095
